Adding lunar coords

wps2n opened the pull request:
Adds compatibility with lunar based observatories using a new keyword `world` to specify the earth or moon (defaults to earth). Also contains a z=30 theory model and zero foregrounds model.

Needs:
Better handling of `time_per_day` and `lst_bin_size`
Moon specific tests

steven-murray commented on ci/testing.yaml:
```diff
@@ -19,3 +19,4 @@ dependencies:
   - matplotlib
   - pip:
     - methodtools
+    - lunarsky>=0.2.5
```
In addition to adding the dependency here, it should be added into the `pyproject.toml` as a dependency

steven-murray commented on src/py21cmsense/_utils.py:
```diff
@@ -39,7 +41,7 @@ def find_nearest(array, value):
 @un.quantity_input
 def phase_past_zenith(
-    time_past_zenith: un.day, bls_enu: np.ndarray, latitude, use_apparent: bool = True
+    time_past_zenith: un.day, bls_enu: np.ndarray, latitude, world, use_apparent: bool = True
```
please give the type hint (`str` I suppose)

steven-murray replied:
and probably give it a default "earth"

steven-murray commented on src/py21cmsense/observation.py:
```diff
@@ -94,16 +94,16 @@ class Observation:
     time_per_day: tp.Time = attr.ib(
         6 * un.hour,
-        validator=(tp.vld_physical_type("time"), ut.between(0 * un.hour, 24 * un.hour)),
+        validator=(tp.vld_physical_type("time"), ut.between(0 * un.hour, 655.2 * un.hour)),
```
Instead of changing the validator here, you could use a more dynamic validator. To do this, remove the `ut.between()` validator from this tuple, and add the following:

```python
@time_per_day.validator
def _time_per_day_vld(self, att, val):
    day_length = 24*un.hour if self.observatory.world == 'earth' else 655.2*un.hour

    if not 0*un.hour <= val <= day_length:
        raise ValueError(f"time_per_day should be between 0 and {day_length}")
```

Note that since the `observatory` attribute is defined before the `time_per_day` attribute on the `Observation` class, it will be available at validation time. Use this same pattern for the other attributes that require adjusting in the PR.

steven-murray commented on src/py21cmsense/sensitivity.py:
```diff
@@ -151,7 +151,7 @@ class PowerSpectrum(Sensitivity):
     horizon_buffer: tp.Wavenumber = attr.ib(default=0.1 * littleh / un.Mpc)
     foreground_model: str = attr.ib(
-        default="moderate", validator=vld.in_(["moderate", "optimistic"])
+        default="moderate", validator=vld.in_(["moderate", "optimistic", "ultra_optimistic"])
```
maybe "foreground-free" is a better name?

steven-murray reviewed (commented):
Thank you very much @wps2n, this is a really cool addition! 

Please see my few notes on the implementation -- otherwise this is very nice indeed :-)

wps2n commented:
@steven-murray I've implemented your suggestions and updated the tests for coverage. Note, I made two pushes. There's an odd error happening with lunarsky in one of the test suites that didn't happen in the first push but now is despite not making any changes to the specific test in the second push. I'll look into that but if you want to review the other changes I think we're nearly ready to merge.

steven-murray reviewed (approved):
Thanks @wps2n this looks great! Let's get those tests passing and I think this is good to go!

wps2n commented:
@steven-murray I passed the PR by Jonathan and made a couple final updates. I am not sure how you would prefer it be merged into main or if there's a couple more housekeeping steps in order. Let me know and I can send it through, or I suppose you can

steven-murray commented:
Great! Thanks @wps2n
